softpack_builder/config/settings.py

```diff
@@ -0,0 +1,145 @@
+"""Copyright (c) Wellcome Sanger Institute.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+"""
+
+import sys
+from pathlib import Path
+from typing import Any, Tuple
+
+import hvac
+import yaml
+from pydantic import BaseSettings
+
+# from pydantic import AnyHttpUrl, AnyUrl, BaseModel, BaseSettings, HttpUrl
+from pydantic.env_settings import SettingsSourceCallable
+
+from softpack_builder.serializable import Serializable
+
+from .models import (
+    ArtifactsConfig,
+    ContainerConfig,
+    EnvironmentsConfig,
+    LoggingConfig,
+    ModulesConfig,
+    ServerConfig,
+    SpackConfig,
+    VaultConfig,
+)
+
+
+class Settings(BaseSettings, Serializable):
+    """Package settings."""
+
+    debug: bool = False
+    server: ServerConfig
+    vault: VaultConfig
+    logging: LoggingConfig
+    spack: SpackConfig
+    environments: EnvironmentsConfig
+    modules: ModulesConfig
+    container: ContainerConfig
+    artifacts: ArtifactsConfig
+
+    class Config:
+        """Configuration loader."""
+
+        config_dir = "conf"
+        config_file = "config.yml"
+
+        @classmethod
+        def file_settings(
+            cls, path: Path, settings: BaseSettings
+        ) -> dict[str, Any]:
+            """Load settings from file.
+
+            Args:
+                path: Config file path.
+                settings: Base settings object.
+
+            Returns:
+                dict[str, Any]: A dictionary of settings.
+            """
+            if not path.is_file():
+                return {}
+            with open(path) as f:
+                return yaml.safe_load(f)
+
+        @classmethod
+        def defaults(cls, settings: BaseSettings) -> dict[str, Any]:
+            """Load defaults from config file.
+
+            Args:
+                settings: BaseSettings model.
+
+            Returns:
+               dict[str, Any]: Settings loaded from default config file.
+            """
+            package_dir = Path(__file__).parent.absolute()
+            path = package_dir / cls.config_dir / cls.config_file
+            return cls.file_settings(path, settings)
```

Comment:
Not sure if this matters, but for GoLang I was told to call functions in the order your program calls them, so that you can read from top to bottom without needing to go up and down repeatedly. 
So with this, it would be,

1. `customise_sources` (nothing calls it)
2. `overrides` (called by `customise_sources`)
3. `file_settings` (called by `overrides`)
4. `vault` (called by `overrides`)
5. `defaults` (called by 'customise_sources')

Reply:
in general you want to organise your code based on some logical grouping and importance, not on calling order so that's just poor advice. For example, consider a File class with a number of operations, open, close, create, delete, read, write, etc.

You want to declare the methods by grouping them based on functionality like this:

```
class File:
    # open and close methods
    open
    close
    
    # create and delete methods
    create
    delete
    
    # read and write methods
    read
    write
```